tsdb/agent/storage.go

```diff
@@ -0,0 +1,678 @@
+				return []record.RefSeries{}
+			},
+		}
+		samplesPool = sync.Pool{
+			New: func() interface{} {
+				return []record.RefSample{}
+			},
+		}
+	)
+
+	go func() {
+		defer close(decoded)
+		for r.Next() {
+			rec := r.Record()
+			switch dec.Type(rec) {
+			case record.Series:
+				series := seriesPool.Get().([]record.RefSeries)[:0]
+				series, err = dec.Series(rec, series)
+				if err != nil {
+					errCh <- &wal.CorruptionErr{
+						Err:     errors.Wrap(err, "decode series"),
+						Segment: r.Segment(),
+						Offset:  r.Offset(),
+					}
+					return
+				}
+				decoded <- series
+			case record.Samples:
+				samples := samplesPool.Get().([]record.RefSample)[:0]
+				samples, err = dec.Samples(rec, samples)
+				if err != nil {
+					errCh <- &wal.CorruptionErr{
+						Err:     errors.Wrap(err, "decode samples"),
+						Segment: r.Segment(),
+						Offset:  r.Offset(),
+					}
+					return
+				}
+				decoded <- samples
+			case record.Tombstones:
+				// We don't care about tombstones
+				continue
+			default:
+				errCh <- &wal.CorruptionErr{
+					Err:     errors.Errorf("invalid record type %v", dec.Type(rec)),
+					Segment: r.Segment(),
+					Offset:  r.Offset(),
+				}
+			}
+		}
+	}()
+
+	for d := range decoded {
+		switch v := d.(type) {
+		case []record.RefSeries:
+			for _, entry := range v {
+				// If this is a new series, create it in memory. If we never read in a
+				// sample for this series, its timestamp will remain at 0 and it will
+				// be deleted at the next GC.
+				if s.series.GetByID(entry.Ref) == nil {
+					series := &memSeries{ref: entry.Ref, lset: entry.Labels, lastTs: 0}
+					s.series.Set(entry.Labels.Hash(), series)
+
+					s.metrics.numActiveSeries.Inc()
+
+					if entry.Ref > lastRef {
+						lastRef = entry.Ref
+					}
+				}
+			}
+
+			//nolint:staticcheck
+			seriesPool.Put(v)
+		case []record.RefSample:
+			for _, entry := range v {
+				// Update the lastTs for the series based
+				series := s.series.GetByID(entry.Ref)
+				if series == nil {
+					level.Warn(s.logger).Log("msg", "found sample referencing non-existing series, skipping")
+					continue
+				}
+
+				series.Lock()
+				if entry.T > series.lastTs {
+					series.lastTs = entry.T
+				}
+				series.Unlock()
+			}
+
+			//nolint:staticcheck
+			samplesPool.Put(v)
+		default:
+			panic(fmt.Errorf("unexpected decoded type: %T", d))
+		}
+	}
+
+	s.nextRef.Store(lastRef)
+
+	select {
+	case err := <-errCh:
+		return err
+	default:
+		if r.Err() != nil {
+			return errors.Wrap(r.Err(), "read records")
+		}
+		return nil
+	}
+}
+
+func (s *Storage) run() {
+	defer close(s.donec)
+
+Loop:
+	for {
+		select {
+		case <-s.stopc:
+			break Loop
+		case <-time.After(s.opts.TruncateFrequency):
+			ts := s.rs.LowestSentTimestamp() - s.opts.MinWALTime.Milliseconds()
+			if ts < 0 {
+				ts = 0
+			}
+
+			// Clamp timestamp to be the maximum lifetime of WAL data.
+			if maxTS := timestamp.FromTime(time.Now().Add(-s.opts.MaxWALTime)); ts < maxTS {
```

Comment:
Above we get a time relative to `LowestSentTimestamp`, and next we get a time relative to `time.Now()`, and I am not able to correlate both. Could you please explain?

That said, the agent (and hence the TSDB) should not use the wall clock to compare the timestamps. It should always be something relative to the some timestamp observed from samples.